lib/di/interceptor/app_interceptor.dart

```diff
@@ -1,26 +1,77 @@
+import 'dart:io';
+
 import 'package:dio/dio.dart';
+import 'package:flutter_survey/di/di.dart';
 import 'package:flutter_survey/local/shared_preference_helper.dart';
+import 'package:flutter_survey/models/oauth_token.dart';
+import 'package:flutter_survey/usecase/base/base_use_case.dart';
+import 'package:flutter_survey/usecase/refresh_token_use_case.dart';
 
-const String HEADER_AUTHORIZATION = 'Authorization';
+const String _headerAuthorization = 'Authorization';
 
-class AppInterceptor extends Interceptor {
-  bool _requireAuthenticate;
-  SharedPreferencesHelper _sharedPreferencesHelper;
+class AppInterceptor extends QueuedInterceptor {
+  final bool _requireAuthenticate;
+  final SharedPreferencesHelper _sharedPreferencesHelper;
+  final Dio _dio;
 
   AppInterceptor(
     this._requireAuthenticate,
     this._sharedPreferencesHelper,
+    this._dio,
   );
 
   @override
-  Future onRequest(
+  void onRequest(
       RequestOptions options, RequestInterceptorHandler handler) async {
     if (_requireAuthenticate) {
-      String accessToken = await _sharedPreferencesHelper.getAccessToken();
-      String tokenType = await _sharedPreferencesHelper.getTokenType();
-      options.headers
-          .putIfAbsent(HEADER_AUTHORIZATION, () => "$tokenType $accessToken");
+      final token = await _buildToken();
+      options.headers.putIfAbsent(_headerAuthorization, () => token);
+    }
+    super.onRequest(options, handler);
+  }
+
+  @override
+  void onError(DioError err, ErrorInterceptorHandler handler) async {
+    final statusCode = err.response?.statusCode;
+    if ((statusCode == HttpStatus.forbidden ||
+            statusCode == HttpStatus.unauthorized) &&
+        _requireAuthenticate) {
+      try {
+        final refreshTokenUseCase = getIt<RefreshTokenUseCase>();
+        final result = await refreshTokenUseCase.call();
+        if (result is Success<OAuthToken>) {
+          // Update new token header
+          final newToken = await _buildToken();
+          err.requestOptions.headers[_headerAuthorization] = newToken;
+
+          // Create request with new access token
+          final options = Options(
+              method: err.requestOptions.method,
+              headers: err.requestOptions.headers);
+          final newRequest = await _dio.request(
+              "${err.requestOptions.baseUrl}${err.requestOptions.path}",
+              options: options,
+              data: err.requestOptions.data,
+              queryParameters: err.requestOptions.queryParameters);
+          handler.resolve(newRequest);
+        } else {
+          handler.next(err);
+        }
+      } catch (exception) {
+        if (exception is DioError) {
+          handler.next(exception);
+        } else {
+          handler.next(err);
+        }
+      }
+    } else {
+      handler.next(err);
```

Comment:
what do you think about this refactoring? 
```dart
final needsRefreshToken = (statusCode == HttpStatus.forbidden || statusCode == HttpStatus.unauthorized) 
      && _requireAuthenticate
if (!needsRefreshToken) { 
    handler.next(err);
    return
}

// refresh token

Reply:
@markgravity I think separating the refresh login into a new method makes more sense. Fixed in https://github.com/luongvo/flutter-survey/pull/78/commits/46de2f9baa246b0ce9b4000afb69dd1dd8373dd2